### ISSUE TITLE
Relax CA2219 in case of a rethrow

### ISSUE THREAD

**Issue by Evangelink** [dotnet/roslyn-analyzers#5869]
### Analyzer

**Diagnostic ID**: [CA2219](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2219)

### Describe the improvement

I'd like to suggest to relax the rule to not report in case of a rethrow inside a try/catch within the finally clause.

```
public void M()
{
	try
	{
		M1();
	}
	finally
	{
		M2();
		try
		{
			MethodThatMightThrow();
		}
		catch (Exception ex)
		{
			// logging
			throw; // Do not raise here as we are not really throwing but rather let an exception flow.
		}
	}
}

### Describe suggestions on how to achieve the rule

Do not report if the exception being thrown in the finally clause is a rethrow.

**Comment by gologames** [comment]
Hello,

I want to up this topic. Agree that rethrow in finally might be a normal exception flow. For example:

`
//Open connection
try
{
	//Write data to connection
}
finally
{
	try
	{
		//Close connection
	}
	catch (Exception ex)
	{
		//Report detailed issue on closing connection
		throw; //Rethrow exception to cancel further execution flow
	}
}
`

**Comment by airbreather** [comment]
Regarding those code samples above on this thread so far: those are cases when I **WOULD** want to see CA2219 get raised.

While I think it would be inappropriate to raise CA2219 in every single `finally` block that *might* throw an exception depending on what external code does, I **DO** think that if a `finally` block is explicitly catching and rethrowing an exception without other considerations, then CA2219 is correct: you've gone out of your way to tell the compiler that you anticipate exceptions in the `finally` block sometimes, and so if you rethrow it, then the original (presumably more important) exception will vanish. I believe that when this happens, most of the time you **really** want that original exception, not the one raised in the `finally` block.

However, I still think that there is actually a reason to relax CA2219 in the case of a certain kind of "rethrow" (of sorts):
```csharp
Exception? primaryException = null;
try
{
    DoMainThingThatMightThrow();
}
catch (Exception e)
{
    primaryException = e;
    throw;
}
finally
{
    try
    {
        DoCleanupThingThatMightAlsoThrow();
    }
    catch (Exception e) when (primaryException is not null)
    {
        // this line raises CA2219:
        throw new AggregateException(primaryException, e);
    }
}
```

This probably looks unusual, but that's precisely **because** it's solving the one problem that CA2219 tries to prevent in the first place: losing the original exception details because a new exception is thrown by some code in the `finally` block. It goes a step further to make sure that not only do the primary exception details get preserved, but that we **also** get the details of the exception that some other external code throws within the `finally` block.

It's ironic, then CA2219 is raised in the code snippet above, but **not** in the following more straightforward bit of code that I believe is basically equivalent to the above **except** for the underlying issue that CA2219 tries to defend against:
```csharp
try
{
    DoMainThingThatMightThrow();
}
finally
{
    // this line does not raise CA2219:
    DoCleanupThingThatMightAlsoThrow();
}
```

IMO this is low-priority because it's pretty clearly a reason for a suppression... though having said that, perhaps updating [the "When to suppress warnings" documentation](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2219#when-to-suppress-warnings) could be some low-hanging fruit that could come out of my comment? 🤷.
